datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -198,68 +211,127 @@ impl ArrayStaticFilter {
+                        return Ok(downcast_array(result.as_ref()))
+                    }
+                    _ => {}
+                }
 
-    fn contains(&self, v: &dyn Array, negated: bool) -> Result<BooleanArray> {
-        let v = v
-            .as_primitive_opt::<Int32Type>()
-            .ok_or_else(|| exec_datafusion_err!("Failed to downcast array"))?;
-
-        let result = match (v.null_count() > 0, negated) {
-            (true, false) => {
-                // has nulls, not negated"
-                BooleanArray::from_iter(
-                    v.iter().map(|value| Some(self.values.contains(&value?))),
-                )
-            }
-            (true, true) => {
-                // has nulls, negated
-                BooleanArray::from_iter(
-                    v.iter().map(|value| Some(!self.values.contains(&value?))),
-                )
-            }
-            (false, false) => {
-                //no null, not negated
-                BooleanArray::from_iter(
-                    v.values().iter().map(|value| self.values.contains(value)),
-                )
-            }
-            (false, true) => {
-                // no null, negated
-                BooleanArray::from_iter(
-                    v.values().iter().map(|value| !self.values.contains(value)),
-                )
+                let v = v
+                    .as_primitive_opt::<$ArrowType>()
+                    .ok_or_else(|| exec_datafusion_err!("Failed to downcast an array to a '{}' array", stringify!($ArrowType)))?;
+
+                let haystack_has_nulls = self.null_count > 0;
+
+                let result = match (v.null_count() > 0, haystack_has_nulls, negated) {
+                    (true, _, false) | (false, true, false) => {
+                        // Either needle or haystack has nulls, not negated
+                        BooleanArray::from_iter(v.iter().map(|value| {
+                            match value {
+                                // SQL three-valued logic: null IN (...) is always null
+                                None => None,
+                                Some(v) => {
+                                    if self.values.contains(&v) {
+                                        Some(true)
+                                    } else if haystack_has_nulls {
+                                        // value not in set, but set has nulls -> null
+                                        None
+                                    } else {
+                                        Some(false)
+                                    }
+                                }
+                            }
+                        }))
+                    }
+                    (true, _, true) | (false, true, true) => {
+                        // Either needle or haystack has nulls, negated
+                        BooleanArray::from_iter(v.iter().map(|value| {
```

Comment:
It probably would be faster to handle the nulls separately or using `set_indices` rather than using `BooleanArray::from_iter` and `v.iter` etc.